C++ interface fixes: reference argument types

GregTheMadMonk opened the pull request:
Related issue: #1904

There are still some fixes that need to be implemented (autodiff on a function pointer doesn't work for some reason, I'll write about it in the issue) as well as tests. Putting it here before these are finished to make sure that my assumptions about how `Duplicated` and `DuplicatedNoNeed` should work were correct and they are indeed only used with reference and pointer arguments.

* Fix move constructors
* Explicitly `impl::tuple_cat` instead of `std::tuple_cat`
* Restrict `Duplicated` and `DuplicatedNoNeed` template arguments to only allow reference and pointer types
* Properly initialize shadow in `Duplicated` and `DuplicatedNoNeed`

GregTheMadMonk commented:
```
my assumptions about how Duplicated and DuplicatedNoNeed should work were correct
```
I've just realized that they are wrong. I assumed that `Duplicated[NoNeed]` only need to work with reference/pointer arguments, but now I realize that regular values my need to be passed in forward mode.

Should there maybe be another class made, maybe `enzyme::ForwardActive`, to stress the difference between reverse mode storing the result in the shadow parameter and forward mode using the shadow parameter as a "direction" to compute the gradient?

wsmoses commented:
@GregTheMadMonk Duplicated / DuplicatedNoNeed permit arbitrary types in forward mode (and also you may want to do a duplicated<int> if say that represents mutable memory in some way e.g. ptrtoint.

It may be worthwhile splitting Duplicated -> Dual in forward mode and we can then add the duplicated restriction in reverse mode -- but that should likely come in a later PR.

wsmoses commented:
@GregTheMadMonk bumping this, could you remove the duplicated error for now [and add a test for reference]

GregTheMadMonk commented:
Yes, sorry, I've encountered a weird issue when trying to make it work with lambdas where I somehow managed to modify the headers so that Enzyme doesn't (seemingly) even start a pass and leaves `__enzyme_autodiff` call as is and then some other stuff kept me distracted for a while. Finishing it rn, so next week (or even tomorrow maybe) I can isolate only duplicated args part and update the PR.

GregTheMadMonk commented:
Sorry for the delay, here is the code.

The code from #1904 is in `gh_issue_1904.cpp` - the test fails only with `-O0`, I've left it in so that it can be looked into later. All other optimizer levels work fine, `dup_args.cpp` passes on all levels.

I've kept the restriction on pointer/reference types only in `Duplicated`/`DuplicatedNoNeed`, but only for reverse mode, and I've put it in `static_assert` with an explaining message.

Some other stuff about C++ interface that I'll be looking into next:
* ~Should Active/Const reference type arguments also get an update?~ Yes, I've just realized they should. I'll add commits here while
* How to make lambdas/other function-like objects work (they obviously are intended to work, but don't as far as I can tell)

wsmoses commented on enzyme/include/enzyme/utils:
```diff
@@ -43,37 +43,67 @@ namespace enzyme {
+            DupArg(DupArg<T&>&&) = default;
+
+            __attribute__((always_inline))
+            const T* valueExp() const { return &this->value; }
```
This shouldn't be const here, no?

wsmoses replied:
It you want to keep const correctness here, maybe we should store T* for value and shadow?

GregTheMadMonk replied:
Yeah `T* value` would be much simpler as it would do the same thing that `valueExp` and `shadowExp` are doing anyway...

As for the `const`... I believe it doesn't matter a lot at this point where the `const` is because the next destination for these pointers is `__enzyme_autodiff` and by the time it reaches IR the semantic value of `const` will be already lost...

wsmoses replied:
That's true, but if Duplicated gets used semantically for how we define custom c++ derivative rules we want users to be able to modify things appropriately.

GregTheMadMonk replied:
That would be a bit tricky to do. Let's take `enzyme::Duplicated<const float&>` for example. What should it's `shadow` type be? In reverse mode it will get modified by the AD, so its const-correct type is `float&`. On the contrary, forward mode will only read the value from that address, so the user should be able to bind a `const` to that reference, which means that the shadow in this case needs to be a `const float&`.

Unless I'm missing something here, there is a contradiction here in what `Reverse` and `Forward` mode want from the shadow parameter in `Duplicated` which can't be solved unless there is a separate class for the `Forward` mode specifically

wsmoses replied:
Yeah we really should fork off duplicated in forward mode to just be Dual and have the semantic changes.

In any case yeah, we may need to remove the const from the primal type in the shadow.

This doesn’t need to stop the PR from getting in to begin with (already improving tuings) — but something to follow up on

wsmoses commented on enzyme/test/Integration/CppSugar/dup_args.cpp:
```diff
@@ -0,0 +1,121 @@
+        const float dfdx = enzyme::get<0>(
+            enzyme::get<0>(
+                enzyme::autodiff<enzyme::Reverse>(
+                    g, enzyme::Active<const float&>{ x }
```
So here I think it would be easiest to just disallow reference active's

GregTheMadMonk replied:
But `Const`s should have a reference specification like `Duplicated`/`DuplicatedNoNeed`, right?

wsmoses replied:
Yup!

wsmoses commented on enzyme/test/Integration/CppSugar/dup_args.cpp:
```diff
@@ -0,0 +1,121 @@
+    }
+#endif
+#if 0
+    {
```
We should be able to make a separate file and do verify errors here. See this for example: https://github.com/EnzymeAD/Enzyme/blob/main/enzyme/test/Integration/ReverseMode/err_empty.c

wsmoses commented:
@GregTheMadMonk once tests pass [and if you can fork off the error checks like above], this should be good to merge!

GregTheMadMonk commented:
* Disabled reference/pointer `Active` specializations via a `static_assert`
* Added a reference-type specialization to `Const`
* Removed `valueExp`/`shadowExp` in favor of just using pointer members
* Removed `is_duplicated` in favor of extending `type_info` (also fixed `DuplicatedNoNeed` not having that trait)
* Added tests to changes
* Added tests that are supposed to be a compile-time error (is there a way to make them not break every time line numbers in enzyme/utils change?)

About `gh_issue_1904`: I'm not sure what to do with it. It just segfaults with `-O0` and it's not a C++-interface problem (checked with `__enzyme_autodiff` and it is the same). I could just remove the `-O0` tests but it would feel like cheating...

jandrej commented:
Sharing the failing code through compiler explorer https://fwd.gymni.ch/PE2qjx. This still doesn't give the correct answer.

wsmoses commented:
@GregTheMadMonk I've fixed an O0 issue with https://github.com/EnzymeAD/Enzyme/pull/1993. Let's see if that makes this succeed.
